Add purge orphan tool

jrafanie opened the pull request:
* Add purge for non-polymorphic orphans
* Add start of a script to purge orphaned rows in known problem models
* Add purge mode support: counts or purge (defaults to counts)

## Example CLI usage:


### Help menu

```
% be ruby tools/purge_orphans.rb -h
** override_gem("byebug", :require=>"byebug") at /Users/joerafaniello/.bundler.d/byebug.rb:1
Usage: purge_orphans [options]
Purge orphaned rows from several tables.

Options:
  -m, --mode=<s>    Mode (default: count) (permitted: count, purge)
  -h, --help        Show this message
```


### Script will default to count if no mode specified

```
% be ruby tools/purge_orphans.rb
** override_gem("byebug", :require=>"byebug") at /Users/joerafaniello/.bundler.d/byebug.rb:1
** ManageIQ master, codename: Tal

Note: Count mode was provided.  No changes are being made.  Use --mode purge to purge the rows.

Purging BinaryBlob...Done!
Purging ContainerCondition...Done!
Purging ContainerEnvVar...Done!
Purging ContainerPortConfig...Done!
Purging ContainerVolume...Done!
Purging CustomAttribute...Done!
Purging MiqReportResultDetail...Done!
Purging RequestLog...Done!
Purging SecurityContext...Done!

Results:

 Class                 | Orphan Count
-----------------------|--------------
 BinaryBlob            |           22
 ContainerCondition    |         2584
 ContainerEnvVar       |          520
 ContainerPortConfig   |          546
 ContainerVolume       |         1876
 CustomAttribute       |         3674
 MiqReportResultDetail |          252
 RequestLog            |         3556
 SecurityContext       |          646

Timing by class:

 Model                 | Time (s)
-----------------------|----------
 BinaryBlob            | 0.089011
 ContainerCondition    | 0.043033
 ContainerEnvVar       | 0.014842
 ContainerPortConfig   | 0.001715
 ContainerVolume       | 0.039398
 CustomAttribute       | 0.144239
 MiqReportResultDetail | 0.046786
 RequestLog            | 0.013888
 SecurityContext       | 0.002554
 TotalTime             | 0.395581
```


### Invalid mode handling

```
% be ruby tools/purge_orphans.rb -m purged
** override_gem("byebug", :require=>"byebug") at /Users/joerafaniello/.bundler.d/byebug.rb:1
Error: option '-m' only accepts one of: count, purge.
Try --help for help.
```

### Purge mode must be specified

```
% be ruby tools/purge_orphans.rb -m purge
** override_gem("byebug", :require=>"byebug") at /Users/joerafaniello/.bundler.d/byebug.rb:1
** ManageIQ master, codename: Tal

Purging BinaryBlob...Done!
Purging ContainerCondition...Done!
Purging ContainerEnvVar...Done!
Purging ContainerPortConfig...Done!
Purging ContainerVolume...Done!
Purging CustomAttribute...Done!
Purging MiqReportResultDetail...Done!
Purging RequestLog...Done!
Purging SecurityContext...Done!

Results:

 Class                 | Purged Rows
-----------------------|-------------
 BinaryBlob            |          22
 ContainerCondition    |        2584
 ContainerEnvVar       |         520
 ContainerPortConfig   |         546
 ContainerVolume       |        1876
 CustomAttribute       |        3674
 MiqReportResultDetail |         252
 RequestLog            |        3556
 SecurityContext       |         646

Timing by class:

 Model                 | Time (s)
-----------------------|----------
 BinaryBlob            | 0.116072
 ContainerCondition    | 0.050117
 ContainerEnvVar       | 0.015245
 ContainerPortConfig   | 0.006876
 ContainerVolume       | 0.053625
 CustomAttribute       | 0.174056
 MiqReportResultDetail | 0.066558
 RequestLog            | 0.029409
 SecurityContext       | 0.007552
 TotalTime             | 0.519666
```


Note, I tested this locally using a neat transaction trick from rails so I could do count or purge mode and not have to restore my database afterwards.

```diff
diff --git a/tools/purge_orphans.rb b/tools/purge_orphans.rb
index 071ec3e513..4798b4b6cf 100755
--- a/tools/purge_orphans.rb
+++ b/tools/purge_orphans.rb
@@ -11,6 +11,7 @@ purge_mode = opts[:mode].to_sym

 # now load rails
 require File.expand_path('../config/environment', __dir__)
+require "active_record/railties/console_sandbox"

 def purge_by_orphaned(klass, fk, window, purge_mode = false)
   klass.include PurgingMixin
```

Fryguy commented on tools/purge_orphans.rb:
```diff
@@ -0,0 +1,35 @@
+def purge_by_orphaned(klass, fk, window)
+  klass.include PurgingMixin
+  klass.define_method(:purge_method) { :destroy }
+  klass.purge_by_orphaned(fk, window).tap {|total| puts "Purged: #{klass}: #{total}"}
```
Minor style
```suggestion
  klass.purge_by_orphaned(fk, window).tap { |total| puts "Purged: #{klass}: #{total}" }
```

Fryguy commented:
Overall, LGTM.

jrafanie commented on app/models/mixins/purging_mixin.rb:
```diff
@@ -152,6 +161,16 @@ def purge_orphans(fk_name, window)
+      assoc_id = connection.quote_column_name("#{fk_name}_id")
+      q_table_name = connection.quote_table_name(table_name)
+
+      scope = joins("LEFT OUTER JOIN #{resource_table} ON #{q_table_name}.#{assoc_id} = #{resource_table}.id")
+              .where(resource_table => {:id => nil})
```
I agree.

Fryguy commented:
One feature request maybe would be a "dry-run" or a count type option. I'd be ok merging without that and it being done in a follow up, but it might be useful for figuring out a rough time estimate or how bad the problem is.

jrafanie commented:
> One feature request maybe would be a "dry-run" or a count type option. I'd be ok merging without that and it being done in a follow up, but it might be useful for figuring out a rough time estimate or how bad the problem is.

I like that idea.  That is usually what we do anyway with real users so it makes sense to add a dry run.

Fryguy commented on app/models/mixins/purging_mixin.rb:
```diff
@@ -152,6 +159,16 @@ def purge_orphans(fk_name, window)
       total
     end
 
+    def purge_non_polymorphic_orphans(fk_name, window, reflection_klass)
+      resource_table = connection.quote_table_name(reflection_klass.table_name)
+      assoc_id = connection.quote_column_name("#{fk_name}_id")
+      q_table_name = connection.quote_table_name(table_name)
+
+      scope = joins("LEFT OUTER JOIN #{resource_table} ON #{q_table_name}.#{assoc_id} = #{resource_table}.id")
+              .where(resource_table => {:id => nil})
+      purge_in_batches(scope, window)
+    end
```
Something like

```suggestion
    def purge_orphans_base_scope(reflection_klass, assoc_id)
      resource_table = connection.quote_table_name(reflection_klass.table_name)
      q_table_name = connection.quote_table_name(table_name)
      joins("LEFT OUTER JOIN #{resource_table} ON #{q_table_name}.#{assoc_id} = #{resource_table}.id")
        .where(resource_table => {:id => nil})
    end
   
    def purge_non_polymorphic_orphans(fk_name, window, reflection_klass)
      assoc_id = connection.quote_column_name("#{fk_name}_id")
      scope = purge_orphans_base_scope(reflection_klass, assoc_id)
      purge_in_batches(scope, window)
    end
```

and change the other method to something like

```ruby
   def purge_polymorphic_orphans(fk_name, window)
      polymorphic_type_column = "#{fk_name}_type"
      polymorphic_id_column   = connection.quote_column_name("#{fk_name}_id")
      total = 0

      polymorphic_classes(polymorphic_type_column).each do |klass|
        scope = purge_orphans_base_scope(klass, polymorphic_id_column)
                   .where("#{q_table_name}.#{connection.quote_column_name(polymorphic_type_column)} = #{connection.quote(klass.name)}")
        total += purge_in_batches(scope, window)
      end
      total
    end
```

Fryguy replied:
Then again, this might make it harder to read 🤷

jrafanie replied:
yeah, I hear you.  I need to change these methods to return scopes in order to do dry run so maybe things can be simplified after some responsibilities are removed from these methods.

Fryguy commented:
Also, I think dry-run or counts should be the _default_, and the user has to opt-in to anything doing actual purging.

jrafanie commented:
> Oh nice - if you have scopes, count is just summing up the scopes counts, right?

Exactly.  The two scopes methods should be easier to refactor after this is done.

jrafanie commented:
> Also, I think dry-run or counts should be the _default_, and the user has to opt-in to anything doing actual purging.

haha, will need a new name...

`ruby tools/purge_orphans.rb --confirm`
`ruby tools/purge_orphans.rb --yes` 

I was thinking it could be similar to installing packages in various platforms with the `-y` option:


yum/dnf:

```
       -y, --assumeyes
              Automatically answer yes for all questions.
```

APT
```
-y, --yes, --assume-yes
           Automatic yes to prompts; assume "yes" as answer to all prompts and
           run non-interactively. If an undesirable situation, such as
           changing a held package, trying to install a unauthenticated
           package or removing an essential package occurs then apt-get will
           abort. Configuration Item: APT::Get::Assume-Yes.  
```

jrafanie commented:
> Hmmm - In other places, I think we just defaulted --dry-run to true, and so they automatically have the --no-dry-run option. I'd have to take a look at the other tools for inspiration

yeah, I can do although I'm not a fan of the negative.  

> EDIT: In purge_metrics.rb we have a --mode parameter with either count or purge choices and count is the default. The user has to pass --mode purge to do the actual purging.
>I could see counts, purge, and dry_run as 3 modes in this case, with counts being the default, >and you could even make --dry-run internally just be a shortcut for --mode dry_run.

I like the `--mode` option a bit better.  I was treating counts as the same as dry-run.  The transaction trick is neat but it doesn't add much over counts so I'd prefer just `count` and `purge` modes with counts as default.

jrafanie commented:
@Fryguy Ok, pushed, please take a look:
* Added purge mode, defaults to counts
* Added purge_by_orphans test for vim performance states to verify the prior behavior at the mixin side still works.  Namely, that it defaults to purge in the mixin.  counts if provided as the purge_mode will return counts.

Verified the container associated tables are covered in the list + request logs (from https://github.com/ManageIQ/manageiq/issues/23307)

We'll add to the list when we do the remaining orphaned tables from the other purgin_mixin models.

Fryguy commented on tools/purge_orphans.rb:
```diff
@@ -0,0 +1,42 @@
+  nil
+end
+puts "Timing by model:"
+pp bm
```
Minor but this might present nicer:

it does require more_core_extensions tableize method.

```suggestion
puts bm.transform_values { |v| v.round(6) }.to_a.unshift(["Model", "Time (s)"]).tableize
```

```text
 Model                 | Time (s)
-----------------------|----------
 BinaryBlob            | 0.097483
 ContainerCondition    | 0.054963
 ContainerEnvVar       | 0.023365
 ContainerPortConfig   | 0.006162
 ContainerVolume       | 0.042559
 CustomAttribute       | 0.151479
 MiqReportResultDetail | 0.085962
 RequestLog            | 0.047107
 SecurityContext       | 0.009384
 TotalTime             | 0.518582
```

jrafanie replied:
haha, yeah, that looks better!  😻

jrafanie replied:
pushed

Fryguy commented on app/models/mixins/purging_mixin.rb:
```diff
@@ -133,23 +133,38 @@ def purge_by_date_and_orphaned(older_than, fk_name, window = purge_window_size)
+      scopes = reflection.polymorphic? ? polymorphic_orphan_scopes(fk_name) : non_polymorphic_orphan_scopes(fk_name, reflection.klass)
+
+      if purge_mode.to_sym == :count
+        scopes.sum {|s| s.count}
```
```suggestion
        scopes.sum(&:count)
```

jrafanie replied:
pushed

Fryguy commented on app/models/mixins/purging_mixin.rb:
```diff
@@ -133,23 +133,38 @@ def purge_by_date_and_orphaned(older_than, fk_name, window = purge_window_size)
+      if purge_mode.to_sym == :count
+        scopes.sum {|s| s.count}
+      else
+        scopes.sum {|s| purge_in_batches(s, window)}
```
Minor rubocop


```suggestion
        scopes.sum { |s| purge_in_batches(s, window) }
```

jrafanie replied:
this never looks right to my eyes 🤣

jrafanie replied:
pushed

Fryguy commented on tools/purge_orphans.rb:
```diff
@@ -0,0 +1,42 @@
+opts = Optimist.options do
+  banner "Purge orphaned rows from several tables.\n\nUsage: ruby #{$PROGRAM_NAME} [options]\n\nOptions:\n\t"
+  opt :mode, "Mode (#{VALID_MODES.join(", ")})", :default => "count"
+end
```
I think you need a check here that only one of the valid modes were passed. I can't remember if optimist has an enum type yet but, if so, you could use that.

jrafanie replied:
thanks, it's permitted, pushing that

```ruby
opt :mode, "Mode (#{VALID_MODES.join(", ")})", :default => "count", :permitted => VALID_MODES
```

pushed

Fryguy replied:
yes that's it - that was added recently.

Fryguy commented on tools/purge_orphans.rb:
```diff
@@ -0,0 +1,42 @@
+
+VALID_MODES = %w[count purge].freeze
+opts = Optimist.options do
+  banner "Purge orphaned rows from several tables.\n\nUsage: ruby #{$PROGRAM_NAME} [options]\n\nOptions:\n\t"
```
There's a better way to do this with optimist I think but this works and I'm not at my computer to try it out

jrafanie replied:
thanks, it's `permitted`, pushing that

```ruby
opt :mode, "Mode (#{VALID_MODES.join(", ")})", :default => "count", :permitted => VALID_MODES
```

jrafanie replied:
I'm not sure what you mean.

Fryguy replied:
You can use synopsis and usage helpers and don't need to reimplement that...For example:

```suggestion
  synopsis "Purge orphaned rows from several tables."
  usage "[options]"
```

Gives you basically the same thing.

jrafanie replied:
ah, nice... done

Fryguy commented on tools/purge_orphans.rb:
```diff
@@ -0,0 +1,42 @@
+
+_result, bm = Benchmark.realtime_block("TotalTime") do
+  CLASSES_TO_PURGE.each_slice(3) do |klass, fk, window|
+    Benchmark.realtime_block(klass.name) do
```
You might want a puts message here like 

```ruby
puts "Purging #{klass.name}"
```

Reason is that if it's a long running table the user won't know which table is the one currently being operated on.

jrafanie replied:
pushed

Fryguy commented:
All of my comments are kind of presentation.  Overall I approve.

Fryguy commented on tools/purge_orphans.rb:
```diff
@@ -0,0 +1,53 @@
+# now load rails
+require File.expand_path('../config/environment', __dir__)
+
+def purge_by_orphaned(klass, fk, window, purge_mode = false)
```
I don't think false is a valid value here?  Think you can just remove this default value.

jrafanie replied:
done

Fryguy commented on tools/purge_orphans.rb:
```diff
@@ -0,0 +1,53 @@
+    Benchmark.realtime_block(klass.name) do
+      print "Purging #{klass.name}..."
+      purge_by_orphaned(klass, fk, window, purge_mode).tap {|result| results << [klass, result]}
+      print "Done!\n"
```
```suggestion
      puts "Done!"
```

jrafanie replied:
haha, I like symmetry

jrafanie replied:
done

Fryguy commented on tools/purge_orphans.rb:
```diff
@@ -0,0 +1,53 @@
+  CLASSES_TO_PURGE.each_slice(3) do |klass, fk, window|
+    Benchmark.realtime_block(klass.name) do
+      print "Purging #{klass.name}..."
+      purge_by_orphaned(klass, fk, window, purge_mode).tap {|result| results << [klass, result]}
```
Minor rubocop

```suggestion
      purge_by_orphaned(klass, fk, window, purge_mode).tap { |result| results << [klass, result] }
```

(I really wish the bot were working 😭)

jrafanie replied:
it will never look right to me

jrafanie replied:
done

jrafanie commented on tools/purge_orphans.rb:
```diff
@@ -0,0 +1,55 @@
+end
+
+CLASSES_TO_PURGE = [
+  BinaryBlob,            :resource,           100,
```
This one was changed to match the existing setting

jrafanie commented on tools/purge_orphans.rb:
```diff
@@ -0,0 +1,55 @@
+  ContainerVolume,       :parent,            1000,
+  CustomAttribute,       :resource,          1000,
+  MiqReportResultDetail, :miq_report_result, 1000,
+  RequestLog,            :resource,           500,
```
This one was changed to match the existing setting for miq requests

Fryguy commented on tools/purge_orphans.rb:
```diff
@@ -0,0 +1,55 @@
+  synopsis "Purge orphaned rows from several tables."
+  usage "[options]"
+
+  opt :mode, "Mode (#{VALID_MODES.join(", ")})", :default => "count", :permitted => VALID_MODES
```
I just realized that now that you're using permitted you don't need to output the values yourself.
```suggestion
  opt :mode, "Mode", :default => "count", :permitted => VALID_MODES
```

jrafanie replied:
pushed, updated description

Fryguy commented:
Backported to `spassky` in commit e02c03be9ac8bac3c7dc62f2cc1e5e0798fdff5c.

```text
commit e02c03be9ac8bac3c7dc62f2cc1e5e0798fdff5c
Author: Jason Frey <fryguy9@gmail.com>
Date:   Wed Apr 2 13:06:30 2025 -0400

    Merge pull request #23399 from jrafanie/add-purge-orphan-tool
    
    Add purge orphan tool
    
    (cherry picked from commit 203fbe02158874ab41e2ca68ccbc41249e09ae29)
```
